File: sample-rxjava/src/test/java/org/dbtools/sample/model/database/DatabaseBaseManagerTest.java
package org.dbtools.sample.model.database;

import org.dbtools.android.domain.DBToolsTableChangeListener;
import org.dbtools.android.domain.DatabaseTableChange;
import org.dbtools.android.domain.config.TestDatabaseConfig;
import org.dbtools.sample.model.database.main.MainDatabaseManagers;
import org.dbtools.sample.model.database.main.individual.Individual;
import org.dbtools.sample.model.database.main.individual.IndividualManager;
import org.junit.Before;
import org.junit.Test;

import java.io.File;
import java.util.List;

import io.reactivex.functions.Consumer;
import io.reactivex.schedulers.Schedulers;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

public class DatabaseBaseManagerTest {

    private DatabaseManager databaseManager;

    @Before
    public void setUp() throws Exception {
        TestDatabaseConfig databaseConfig = new TestMainDatabaseConfig("java-test-individual.db");
        databaseConfig.deleteAllDatabaseFiles();
        databaseManager = new DatabaseManager(databaseConfig);
        MainDatabaseManagers.init(databaseManager);
    }

    /**
     * @return swap database File
     */
    private File setupSwap() {
        // create a database to swap with
        IndividualManager individualManager = MainDatabaseManagers.getIndividualManager();

        Individual ind1 = new Individual();
        ind1.setFirstName("A");

        individualManager.save(ind1);

        assertEquals(1, individualManager.findCount());
        assertEquals("A", individualManager.findAll().get(0).getFirstName());

        // close the database
        databaseManager.closeDatabase(DatabaseManagerConst.MAIN_DATABASE_NAME);
        String databaseAPath = databaseManager.getDatabase(DatabaseManagerConst.MAIN_DATABASE_NAME).getPath();

        // move the database file to a swap database <database path>2
        File databaseAFile = new File(databaseAPath);
        File swapDatabaseFile = new File(databaseAPath + "2");
        assertTrue(databaseAFile.renameTo(swapDatabaseFile));


        // start to write again... which will create a new database
        Individual ind2 = new Individual();
        ind2.setFirstName("B");
        individualManager.save(ind2);

        Individual ind3 = new Individual();
        ind3.setFirstName("C");
        individualManager.save(ind3);

        // Make sure the new database exists
        List<Individual> individua2lList = individualManager.findAll();
        assertEquals(2, individua2lList.size());
        assertEquals("B", individua2lList.get(0).getFirstName());
        assertEquals("C", individua2lList.get(1).getFirstName());

        return swapDatabaseFile;
    }

    @Test
    public void testSwap() {
        IndividualManager individualManager = MainDatabaseManagers.getIndividualManager();
        File swapDatabaseFile = setupSwap();

        assertTrue(databaseManager.swapDatabase(DatabaseManagerConst.MAIN_DATABASE_NAME, swapDatabaseFile));

        assertEquals(1, individualManager.findCount());
        assertEquals("A", individualManager.findAll().get(0).getFirstName());
    }

    @Test
    public void testSwapInTransaction() {
        IndividualManager individualManager = MainDatabaseManagers.getIndividualManager();
        File swapDatabaseFile = setupSwap();

        individualManager.beginTransaction();

        assertTrue(databaseManager.swapDatabase(DatabaseManagerConst.MAIN_DATABASE_NAME, swapDatabaseFile));

        individualManager.endTransaction(true);

        assertEquals(1, individualManager.findCount());
        assertEquals("A", individualManager.findAll().get(0).getFirstName());
    }

    @Test
    public void testSwapErrors() {
        IndividualManager individualManager = MainDatabaseManagers.getIndividualManager();
        File swapDatabaseFile = setupSwap();

        assertFalse(databaseManager.swapDatabase(DatabaseManagerConst.MAIN_DATABASE_NAME, new File("/data/bad-file.db")));
        assertFalse(databaseManager.swapDatabase("BAD_DATABASE_NAME", swapDatabaseFile));

        // original database should NOT have changed
        List<Individual> individua2lList = individualManager.findAll();
        assertEquals(2, individua2lList.size());
        assertEquals("B", individua2lList.get(0).getFirstName());
        assertEquals("C", individua2lList.get(1).getFirstName());
    }

    private int subscribeMessageCount; // listener
    private int listenerMessageCount; // listener
    @Test
    public void testSwapSubscriptionAndListener() {
        subscribeMessageCount = 0;
        listenerMessageCount = 0;

        IndividualManager individualManager = MainDatabaseManagers.getIndividualManager();
        individualManager.addTableChangeListener(new DBToolsTableChangeListener() {
            @Override
            public void onTableChange(DatabaseTableChange tableChange) {
                System.out.println("Table Changed: " + tableChange.getTable() + " ID: " + tableChange.getRowId());
                listenerMessageCount++;
            }
        });

        individualManager.tableChanges()
                .subscribeOn(Schedulers.trampoline())
                .observeOn(Schedulers.trampoline())
                .subscribe(new Consumer<DatabaseTableChange>() {
                    @Override
                    public void accept(DatabaseTableChange tableChange) {
                        System.out.println("Table Changed: " + tableChange.getTable() + " ID: " + tableChange.getRowId());
                        subscribeMessageCount++;
                    }
                });

        File swapDatabaseFile = setupSwap();

        int lastListenerCount = listenerMessageCount;
        int lastSubscribeCount = subscribeMessageCount;

        assertTrue(databaseManager.swapDatabase(DatabaseManagerConst.MAIN_DATABASE_NAME, swapDatabaseFile));

        Individual newIndividual = new Individual();
        newIndividual.setFirstName("New");
        individualManager.save(newIndividual);

        assertEquals(lastListenerCount + 1, listenerMessageCount);
        assertEquals(lastSubscribeCount + 1, subscribeMessageCount);
    }
}

File: sample-basic-kotlin/src/test/kotlin/org/dbtools/sample/kotlin/model/database/DatabaseBaseManagerTest.kt
package org.dbtools.sample.kotlin.model.database

import org.dbtools.android.domain.DBToolsTableChangeListener
import org.dbtools.sample.kotlin.model.database.main.MainDatabaseManagers
import org.dbtools.sample.kotlin.model.database.main.individual.Individual
import org.junit.Assert.*
import org.junit.Before
import org.junit.Test
import java.io.File

class DatabaseBaseManagerTest {

    private lateinit var databaseManager: DatabaseManager

    @Before
    @Throws(Exception::class)
    fun setUp() {
        val databaseConfig = TestMainDatabaseConfig("java-test-individual.db")
        databaseConfig.deleteAllDatabaseFiles()
        databaseManager = DatabaseManager(databaseConfig)
        MainDatabaseManagers.init(databaseManager)
    }

    /**
     * @return swap database File
     */
    private fun setupSwap(): File {
        // create a database to swap with
        val individualManager = MainDatabaseManagers.individualManager

        val ind1 = Individual()
        ind1.firstName = "A"

        individualManager.save(ind1)

        assertEquals(1, individualManager.findCount())
        assertEquals("A", individualManager.findAll()[0].firstName)

        // close the database
        databaseManager.closeDatabase(DatabaseManagerConst.MAIN_DATABASE_NAME)
        val databaseAPath = databaseManager.getDatabase(DatabaseManagerConst.MAIN_DATABASE_NAME)!!.path

        // move the database file to a swap database <database path>2
        val databaseAFile = File(databaseAPath)
        val swapDatabaseFile = File(databaseAPath + "2")
        assertTrue(databaseAFile.renameTo(swapDatabaseFile))


        // start to write again... which will create a new database
        val ind2 = Individual()
        ind2.firstName = "B"
        individualManager.save(ind2)

        val ind3 = Individual()
        ind3.firstName = "C"
        individualManager.save(ind3)

        // Make sure the new database exists
        val individua2lList = individualManager.findAll()
        assertEquals(2, individua2lList.size.toLong())
        assertEquals("B", individua2lList[0].firstName)
        assertEquals("C", individua2lList[1].firstName)

        return swapDatabaseFile
    }

    @Test
    fun testSwap() {
        val individualManager = MainDatabaseManagers.individualManager
        val swapDatabaseFile = setupSwap()

        assertTrue(databaseManager.swapDatabase(DatabaseManagerConst.MAIN_DATABASE_NAME, swapDatabaseFile))

        assertEquals(1, individualManager.findCount())
        assertEquals("A", individualManager.findAll()[0].firstName)
    }

    @Test
    fun testSwapInTransaction() {
        val individualManager = MainDatabaseManagers.individualManager
        val swapDatabaseFile = setupSwap()

        individualManager.beginTransaction()

        assertTrue(databaseManager.swapDatabase(DatabaseManagerConst.MAIN_DATABASE_NAME, swapDatabaseFile))

        individualManager.endTransaction(true)

        assertEquals(1, individualManager.findCount())
        assertEquals("A", individualManager.findAll()[0].firstName)
    }

    @Test
    fun testSwapErrors() {
        val individualManager = MainDatabaseManagers.individualManager
        val swapDatabaseFile = setupSwap()

        assertFalse(databaseManager.swapDatabase(DatabaseManagerConst.MAIN_DATABASE_NAME, File("/data/bad-file.db")))
        assertFalse(databaseManager.swapDatabase("BAD_DATABASE_NAME", swapDatabaseFile))

        // original database should NOT have changed
        val individua2lList = individualManager.findAll()
        assertEquals(2, individua2lList.size.toLong())
        assertEquals("B", individua2lList[0].firstName)
        assertEquals("C", individua2lList[1].firstName)
    }

    private var listenerMessageCount: Int = 0 // listener
    @Test
    fun testSwapListener() {
        listenerMessageCount = 0

        val individualManager = MainDatabaseManagers.individualManager
        individualManager.addTableChangeListener(DBToolsTableChangeListener { tableChange ->
            println("Table Changed: " + tableChange.table + " ID: " + tableChange.rowId)
            listenerMessageCount++
        })

        val swapDatabaseFile = setupSwap()

        val lastListenerCount = listenerMessageCount

        assertTrue(databaseManager.swapDatabase(DatabaseManagerConst.MAIN_DATABASE_NAME, swapDatabaseFile))

        val newIndividual = Individual()
        newIndividual.firstName = "New"
        individualManager.save(newIndividual)

        assertEquals((lastListenerCount + 1).toLong(), listenerMessageCount.toLong())
    }
}